README.md
## storemeta


- [x] This is a simple golang script that will parse/scrape the apps-store/play-store  meta-infos:

####        a. Android Play Store meta info

######      https://play.google.com/store/apps/details?id=com.google.android.apps.photos&hl=en


####        b. Apple App Store meta info 

######      https://itunes.apple.com/app/id293622097?mt=8

- [x] Output the meta-info in JSON format



## Compile

```sh

     git clone https://github.com/bayugyug/storemeta.git && cd storemeta 

     git pull && make 

```


## Meta-Infos

```go

        type App struct {
                Platform        string `json:"platform"`
                AppURL          string `json:"app-url"`
                AppID           string `json:"app-id"`
                Genre           string `json:"genre"`
                Title           string `json:"title"`
                Description     string `json:"description"`
                Badge           string `json:"badge"`
                RatingTotal     string `json:"rating-total"`
                RatingPerStar   string `json:"rating-per-star"`
                RatingDesc      string `json:"rating-desc"`
                RatingValue     string `json:"rating-value"`
                SoftwarePrice   string `json:"software-price"`
                MetaDesc        string `json:"meta-desc"`
                FileSize        string `json:"file-size"`
                ContentRating   string `json:"content-rating"`
                DatePublished   string `json:"date-published"`
                SoftwareVersion string `json:"software-version"`
                SoftwareOs      string `json:"software-os"`
                TotalDownloads  string `json:"total-downloads"`
                Developer       string `json:"developer"`
                DeveloperSite   string `json:"developer-site"`
        }

```





## Usage

```sh

$ ./storemeta -h

Ver: 0.1.0-20160830.022006

  -a string
        use for querying the Android App Store (shorthand)
  -android string
        use for querying the Android App Store
  -ca string
        use for querying the list of apps per category in IOS App Store (shorthand)
  -category-android string
        use for querying the list of apps per category in IOS App Store
  -category-ios string
        use for querying the list of apps per category in IOS App Store
  -ci string
        use for querying the list of apps per category in IOS App Store (shorthand)
  -d    use to enable the output in console (shorthand) (default true)
  -debug
        use to enable the output in console (default true)
  -h    Show this help/how-to
  -i string
        use for querying the IOS App Store (shorthand)
  -ios string
        use for querying the IOS App Store
  -la
        use for querying the category list of apps in Android App Store (shorthand)
  -li
        use for querying the category list of apps in IOS App Store (shorthand)
  -list-category-android
        use for querying the category list of apps in Android App Store
  -list-category-ios
        use for querying the category list of apps in IOS App Store
  -pf string
        use to enable what format is used in showing the output (shorthand) (default "json")
  -print-format string
        use to enable what format is used in showing the output (default "json")


        Example:

                ./storemeta -a <AndroidStoreID>  -i <IOSStoreID>

                ./storemeta -list-category-android

                ./storemeta -list-category-ios

                ./storemeta -category-android=GAME_ACTION

                ./storemeta -category-ios=GAMES_ACTION

                ./storemeta  -a="com.google.android.apps.photos"

                or

                ./storemeta  -i="293622097"

                or

                ./storemeta  -a="com.google.android.apps.photos" -i="293622097"

                or

                ./storemeta  -a="com.google.android.apps.plus,com.google.android.launcher,com.sphero.sprk"

                or

                ./storemeta  -i="544007664,535886823,643496868"

```


## Sample (Android)



#### Simple
```go


$    ./storemeta  -a="com.google.android.apps.photos"
    {
        "platform": "ANDROID",
        "title": "Google Photos - Android Apps on Google Play",
        "developer": "Google Inc.",
        "developer-site": "https://www.google.com/url?q=http://www.google.com/policies/privacy\u0026sa=D\u0026usg=AFQjCNE7y6nm7TcHvct7CDJRmWrYBHvMEQ",
        "genre": "PHOTOGRAPHY",
        "description": "Google Photos is the home for all your photos and videos. Automatically organized and searchable, you can find photos fast and bring them to life. It’s the photo gallery that thinks like you do.  VISUAL SEARCH Your photos are now searchable by the places and things that appear in them. Looking for that fish taco you ate in Hawaii? Just search “food in Hawaii” to find it – no tagging required.  UNLIMITED FREE HIGH QUALITY STORAGE Automatically backup all your photos and videos. Access them on any device or on the web at photos.google.com. Your photos are safe, secure, and private to you.  FREE UP SPACE ON YOUR DEVICE Never worry about running out of space on your phone again. In Settings, just tap “Free up device storage” – photos that are safely backed up will be removed from your device’s storage, but will still be available in Google Photos.  BRING PHOTOS TO LIFE Enjoy automatically created montage movies, interactive stories, collages, animations, panoramas, and more from your photos. Or you can easily create them yourself – just tap +.  EASY EDITING Transform photos with the tap of a finger. Use simple, yet powerful, photo and video editing tools to apply filters, adjust colors, and more.  SHARED ALBUMS Get everyone’s photos and videos in one place, across Android, iOS, and the web. Privately sharing all the photos you took – and getting the ones you didn’t – has never been easier.  INSTANT SHARING Instantly share up to 1,500 photos with anyone – no matter what device they’re on. In the share menu, just tap Create Link.  REDISCOVER YOUR PHOTOS It’s easier than ever to relive your memories. The Assistant can create collages of your old photos that help you relive the past.  READY TO CAST View your photos and videos on your TV with Chromecast support.  Follow us for the latest news and updates Twitter: https://twitter.com/GooglePhotos Google+: https://google.com/+GooglePhotos  Need help? Visit https://support.google.com/photos  Face grouping is not available in all countries.",
        "badge": "Top Developer",
        "rating-total": "4846787",
        "rating-per-star": "3214313,920228,344084,130721,237441",
        "rating-desc": "Rated 4.4 stars out of five stars\nRated 5 stars out of five stars\nRated 4 stars out of five stars\nRated 2 stars out of five stars\nRated 3 stars out of five stars\nRated 1 stars out of five stars\nRated 4.3 stars out of five stars\nRated 4.5 stars out of five stars\nRated 4.2 stars out of five stars\nRated 4.6 stars out of five stars\nRated 3.9 stars out of five stars\nRated 4.1 stars out of five stars\nRated 4.0 stars out of five stars",
        "rating-value": "4.391282558441162",
        "software-price": "0",
        "meta-desc": "\r\n Updated August 22, 2016 \r\n Installs   500,000,000 - 1,000,000,000   \r\n Requires Android       4.0 and up     \r\n Content Rating Rated for 3+  Learn more  \r\n Permissions  View details  \r\n Report  Flag as inappropriate  \r\n  Offered By  Google Inc. \r\n  Developer    Visit website   Email apps-help@google.com   Privacy Policy  1600 Amphitheatre Parkway, Mountain View 94043  ",
        "file-size": "",
        "content-rating": "Rated for 3+",
        "date-published": "",
        "software-version": "",
        "software-os": "4.0 and up",
        "total-downloads": "750000000",
        "app-url": "https://play.google.com/store/apps/details?id=com.google.android.apps.photos\u0026hl=en",
        "app-id": "com.google.android.apps.photos"
   }
```

## Sample (IOS)

#### Simple
```go


$   ./storemeta  -i="293622097"

    {
        "platform": "IOS",
        "title": "Google Earth on the App StoreGoogle Earth\n              By Google, Inc.\n              \n            \n            \n              \n              View More by This Developer\n              \n            \n            Open iTunes to buy and download apps.",
        "developer": "Google, Inc.",
        "developer-site": "https://itunes.apple.com/WebObjects/MZStore.woa/wa/viewEula?id=293622097",
        "genre": "TRAVEL",
        "description": "Fly around the planet with a swipe of your finger with Google Earth for iPhone, iPad, and iPod touch. Explore distant lands or reacquaint yourself with your childhood home. Search for cities, places, and businesses. Browse layers including roads, borders, places, photos and more.  See the world at street level with integrated Street View.Use the “tour guide” to easily discover exciting new places to explore. With a quick swipe on the tab at the bottom of the screen, you can bring up a selection of virtual tours from around the globe.With 3D imagery, you can now fly through complete 3D recreations of select cities, including San Francisco, Boston, Rome, and others. With every building modeled in 3D, you truly get a sense of flying above the city.",
        "badge": "This app is designed for both iPhone and iPad",
        "rating-total": "446047",
        "rating-per-star": "378,445669",
        "rating-desc": "3 and a half stars, 378 Ratings",
        "rating-value": "3.69312",
        "software-price": "Free",
        "meta-desc": "",
        "file-size": "30.1",
        "content-rating": "Rated 4+",
        "date-published": "2016-05-03 00:00:00",
        "software-version": "7.1.6",
        "software-os": "Requires iOS 5.0 or later. Compatible with iPhone, iPad, and iPod touch.",
        "total-downloads": "44604700",
        "app-url": "https://itunes.apple.com/us/app/google-earth/id293622097?mt=8",
        "app-id": "293622097"
    }
```


```

## List the Apps Categories

#### (IOS)

```go


$   ./storemeta -print-format=simple -list-category-ios
        1 BOOKS
        2 BUSINESS
        3 EDUCATION
        4 ENTERTAINMENT
        5 FINANCE
        6 FOOD_DRINK
        7 GAMES
        8 GAMES_ACTION
        9 GAMES_ADVENTURE
        10 GAMES_ARCADE
        11 GAMES_BOARD
        12 GAMES_CARD
        13 GAMES_CASINO
        14 GAMES_DICE
        15 GAMES_EDUCATIONAL
        16 GAMES_FAMILY
        17 GAMES_MUSIC
        18 GAMES_PUZZLE
        19 GAMES_RACING
        20 GAMES_ROLE_PLAYING
        21 GAMES_SIMULATION
        22 GAMES_SPORTS
        23 GAMES_STRATEGY
        24 GAMES_TRIVIA
        25 GAMES_WORD
        26 HEALTH_FITNESS
        27 LIFESTYLE
        28 MAGAZINES_NEWSPAPERS
        29 MEDICAL
        30 MUSIC
        31 NAVIGATION
        32 NEWS
        33 PHOTO_VIDEO
        34 PRODUCTIVITY
        35 REFERENCE
        36 SHOPPING
        37 SOCIAL_NETWORKING
        38 SPORTS
        39 TRAVEL
        40 UTILITIES
        41 WEATHER

```


#### (ANDROID)

```go

$       ./storemeta  -print-format=simple -list-category-android

        1 ANDROID_WEAR
        2 BOOKS_AND_REFERENCE
        3 BUSINESS
        4 COMICS
        5 COMMUNICATION
        6 EDUCATION
        7 ENTERTAINMENT
        8 FINANCE
        9 HEALTH_AND_FITNESS
        10 LIBRARIES_AND_DEMO
        11 LIFESTYLE
        12 MEDIA_AND_VIDEO
        13 MEDICAL
        14 MUSIC_AND_AUDIO
        15 NEWS_AND_MAGAZINES
        16 PERSONALIZATION
        17 PHOTOGRAPHY
        18 PRODUCTIVITY
        19 SHOPPING
        20 SOCIAL
        21 SPORTS
        22 TOOLS
        23 TRANSPORTATION
        24 TRAVEL_AND_LOCAL
        25 WEATHER
        26 GAME
        27 GAME_ACTION
        28 GAME_ADVENTURE
        29 GAME_ARCADE
        30 GAME_BOARD
        31 GAME_CARD
        32 GAME_CASINO
        33 GAME_CASUAL
        34 GAME_EDUCATIONAL
        35 GAME_MUSIC
        36 GAME_PUZZLE
        37 GAME_RACING
        38 GAME_ROLE_PLAYING
        39 GAME_SIMULATION
        40 GAME_SPORTS
        41 GAME_STRATEGY
        42 GAME_TRIVIA
        43 GAME_WORD
        44 FAMILY
        45 FAMILY_AGE_AGE_RANGE1
        46 FAMILY_AGE_AGE_RANGE2
        47 FAMILY_AGE_AGE_RANGE3
        48 FAMILY_ACTION
        49 FAMILY_BRAINGAMES
        50 FAMILY_CREATE
        51 FAMILY_EDUCATION
        52 FAMILY_MUSICVIDEO
        53 FAMILY_PRETEND



```


## List the App in the Category

#### (IOS)

```go

$ ./storemeta -category-ios=GAMES_ACTION
[        {
                "preview": "https://itunes.apple.com/us/app/call-of-mini-zombies/id431213733?mt=8",
                "store_id": "431213733",
                "category": "GAMES_ACTION"
        },
        {
                "preview": "https://itunes.apple.com/us/app/gun-club-2-best-in-virtual/id311594640?mt=8",
                "store_id": "311594640",
                "category": "GAMES_ACTION"
        },
        {
                "preview": "https://itunes.apple.com/us/app/ace-fishing-wild-catch/id694972182?mt=8",
                "store_id": "694972182",
                "category": "GAMES_ACTION"
        },
        {
                "preview": "https://itunes.apple.com/us/app/respawnables/id575684686?mt=8",
                "store_id": "575684686",
                "category": "GAMES_ACTION"
        },
        {
                "preview": "https://itunes.apple.com/us/app/swing/id1064078609?mt=8",
                "store_id": "1064078609",
                "category": "GAMES_ACTION"
        }
]

```

#### (ANDROID)

```go
$ ./storemeta -category-android=FAMILY_AGE_AGE_RANGE1
[
        {
                "preview": "https://play.google.com/store/apps/details?id=com.disney.PaintAndPlay_goo",
                "store_id": "com.disney.PaintAndPlay_goo",
                "category": "FAMILY_AGE_AGE_RANGE1"
        },
        {
                "preview": "https://play.google.com/store/apps/details?id=com.intellijoy.android.shapes",
                "store_id": "com.intellijoy.android.shapes",
                "category": "FAMILY_AGE_AGE_RANGE1"
        },
        {
                "preview": "https://play.google.com/store/apps/details?id=com.intellijoy.android.shapes",
                "store_id": "com.intellijoy.android.shapes",
                "category": "FAMILY_AGE_AGE_RANGE1"
        },
        {
                "preview": "https://play.google.com/store/apps/details?id=com.intellijoy.android.shapes",
                "store_id": "com.intellijoy.android.shapes",
                "category": "FAMILY_AGE_AGE_RANGE1"
        },
        {
                "preview": "https://play.google.com/store/apps/details?id=com.intellijoy.android.shapes",
                "store_id": "com.intellijoy.android.shapes",
                "category": "FAMILY_AGE_AGE_RANGE1"
        }
]



```

## Note:

- [x] IOS Total Downloads is calculated :

        Rating-Total x 10

- [x] Android Total Downloads is calculated :
    
        ( As Is in "Installs" )


## Docker Binary  (https://registry.hub.docker.com)

- [x] In order to  use it as dockerize binary


```sh

    sudo  sysctl -w net.ipv4.ip_forward=1

    sudo  docker run --rm  registry.hub.docker.com/bayugyug/storemeta -h

```

## Docker Binary  (https://registry.gitlab.com)


```sh

    sudo  sysctl -w net.ipv4.ip_forward=1

    sudo  docker run --rm  registry.gitlab.com/bayugyug/storemeta -h

```

## As HTTP Server

- [x] In order to  use it via CURL/WGET or Browser


```sh

    sudo  sysctl -w net.ipv4.ip_forward=1

    sudo  docker run -p 7000-8000:7000-8000 -v `pwd`:`pwd` -w `pwd` -d --name storemeta-latest  registry.hub.docker.com/bayugyug/storemeta:latest --http --port 7778

    sudo  docker run -p 7000-8000:7000-8000 -v `pwd`:`pwd` -w `pwd` -d --name storemeta-latest  registry.gitlab.com/bayugyug/storemeta:latest --http --port 7778

    curl -i -v 'http://127.0.0.1:7778/list-category-android'

    curl -i -v 'http://127.0.0.1:7778/list-category-ios'

    curl -i -v 'http://127.0.0.1:7778/category-android/?p=GAME_ACTION'

    curl -i -v 'http://127.0.0.1:7778/category-ios/?p=GAMES_ACTION'

    curl -i -v 'http://127.0.0.1:7778/storeid/?a=com.google.android.apps.photos&i=293622097'

```

### License

[MIT](https://bayugyug.mit-license.org/)



CHANGELOG.md
#CHANGELOG

###storemeta 1.0beta0.1 , 2016-08-29

- initial push

###storemeta 1.0beta0.2 , 2016-09-08

- reduce the docker image file size


###storemeta 1.0beta0.3 , 2019-01-05

- Update the parsers for the IOS/ANDROID , since they've change a lot on their templates




global.go
package main

import (
	"crypto/x509"
	"flag"
	"fmt"
	"io"
	"log"
	"math/rand"
	"net/http"
	"os"
	"path"
	"regexp"
	"strings"
	"time"

	"github.com/julienschmidt/httprouter"
)

const (
	usageShowConsole     = "use to enable the output in console"
	usageAndroidStore    = "use for querying the Android App Store"
	usageIOSStore        = "use for querying the IOS App Store"
	usageAndroidList     = "use for querying the category list of apps in Android App Store"
	usageIOSList         = "use for querying the category list of apps in IOS App Store"
	usageAndroidCategory = "use for querying the list of apps per category in Android App Store"
	usageIOSCategory     = "use for querying the list of apps per category in IOS App Store"
	usagePrintFormat     = "use to enable what format is used in showing the output"
	usageTestingConsole  = "use to enable a dummy flag for testing purposes"
	//IOS device
	IOS = "IOS"
	//ANDROID device
	ANDROID       = "ANDROID"
	usageHTTP     = "use to serve the results via HTTP"
	usageHTTPPort = "use to serve the results via HTTP @ Port No."
)

//StoreApp ... data holder for ios data
type StoreApp struct {
	OS       string `json:"os,omitempty"`
	StoreID  string `json:"store_id,omitempty"`
	Preview  string `json:"preview,omitempty"`
	Category string `json:"category,omitempty"`
	Name     string `json:"name,omitempty"`
	URL      string `json:"url,omitempty"`
}

//App ... meta info holder
type App struct {
	Platform        string `json:"platform"`
	AppURL          string `json:"app-url"`
	AppID           string `json:"app-id"`
	Genre           string `json:"genre"`
	Title           string `json:"title"`
	Description     string `json:"description"`
	Badge           string `json:"badge"`
	RatingTotal     string `json:"rating-total"`
	RatingPerStar   string `json:"rating-per-star"`
	RatingDesc      string `json:"rating-desc"`
	RatingValue     string `json:"rating-value"`
	SoftwarePrice   string `json:"software-price"`
	MetaDesc        string `json:"meta-desc"`
	FileSize        string `json:"file-size"`
	ContentRating   string `json:"content-rating"`
	DatePublished   string `json:"date-published"`
	SoftwareVersion string `json:"software-version"`
	SoftwareOs      string `json:"software-os"`
	TotalDownloads  string `json:"total-downloads"`
	Developer       string `json:"developer"`
	DeveloperSite   string `json:"developer-site"`
}

//Formatter ... generic func for different formatting
type Formatter struct {
	Mode   string
	Format func(AppsMeta, string, string) string
}

var (
	pLogDir = "."
	//loggers
	infoLog  *log.Logger
	warnLog  *log.Logger
	errorLog *log.Logger
	//stats
	//pStats *StatsHelper
	//signal flag
	pStillRunning = true

	pBuildTime = "0"
	pVersion   = "0.1.3" + "-" + pBuildTime
	//console
	pShowConsole    = true
	pAndroidStoreID = ""
	pIOSStoreID     = ""

	pIOSList     = false
	pAndroidList = false

	pIOSCategory     = ""
	pAndroidCategory = ""

	//envt
	pEnvVars = map[string]*string{
		"GMONGERS_LDIR": &pLogDir,
	}
	pTestingConsole = false
	//print_format
	pPrintFormat = "json"
	pHelp        = false

	pStoreURI = map[string][]string{
		ANDROID: {"https://play.google.com/store/apps/details?id=", "&hl=en"},
		IOS:     {"https://itunes.apple.com/app/id", "?mt=8"},
	}
	//ssl certs
	pool *x509.CertPool

	pAppList []*App

	pHTTPPort  = "7777"
	pHTTPServe = false

	//Formatters ... mapping of all the format funcs
	Formatters map[string]Formatter
	pAppsData  chan *App
	pStores    []*StoreApp
)

type logOverride struct {
	Prefix string `json:"prefix,omitempty"`
}

var pCategories = map[string][]string{
	ANDROID: {
		"https://play.google.com/store/apps/category/ANDROID_WEAR",
		"https://play.google.com/store/apps/category/BOOKS_AND_REFERENCE",
		"https://play.google.com/store/apps/category/BUSINESS",
		"https://play.google.com/store/apps/category/COMICS",
		"https://play.google.com/store/apps/category/COMMUNICATION",
		"https://play.google.com/store/apps/category/EDUCATION",
		"https://play.google.com/store/apps/category/ENTERTAINMENT",
		"https://play.google.com/store/apps/category/FINANCE",
		"https://play.google.com/store/apps/category/HEALTH_AND_FITNESS",
		"https://play.google.com/store/apps/category/LIBRARIES_AND_DEMO",
		"https://play.google.com/store/apps/category/LIFESTYLE",
		"https://play.google.com/store/apps/category/MEDIA_AND_VIDEO",
		"https://play.google.com/store/apps/category/MEDICAL",
		"https://play.google.com/store/apps/category/MUSIC_AND_AUDIO",
		"https://play.google.com/store/apps/category/NEWS_AND_MAGAZINES",
		"https://play.google.com/store/apps/category/PERSONALIZATION",
		"https://play.google.com/store/apps/category/PHOTOGRAPHY",
		"https://play.google.com/store/apps/category/PRODUCTIVITY",
		"https://play.google.com/store/apps/category/SHOPPING",
		"https://play.google.com/store/apps/category/SOCIAL",
		"https://play.google.com/store/apps/category/SPORTS",
		"https://play.google.com/store/apps/category/TOOLS",
		"https://play.google.com/store/apps/category/TRANSPORTATION",
		"https://play.google.com/store/apps/category/TRAVEL_AND_LOCAL",
		"https://play.google.com/store/apps/category/WEATHER",
		"https://play.google.com/store/apps/category/GAME",
		"https://play.google.com/store/apps/category/GAME_ACTION",
		"https://play.google.com/store/apps/category/GAME_ADVENTURE",
		"https://play.google.com/store/apps/category/GAME_ARCADE",
		"https://play.google.com/store/apps/category/GAME_BOARD",
		"https://play.google.com/store/apps/category/GAME_CARD",
		"https://play.google.com/store/apps/category/GAME_CASINO",
		"https://play.google.com/store/apps/category/GAME_CASUAL",
		"https://play.google.com/store/apps/category/GAME_EDUCATIONAL",
		"https://play.google.com/store/apps/category/GAME_MUSIC",
		"https://play.google.com/store/apps/category/GAME_PUZZLE",
		"https://play.google.com/store/apps/category/GAME_RACING",
		"https://play.google.com/store/apps/category/GAME_ROLE_PLAYING",
		"https://play.google.com/store/apps/category/GAME_SIMULATION",
		"https://play.google.com/store/apps/category/GAME_SPORTS",
		"https://play.google.com/store/apps/category/GAME_STRATEGY",
		"https://play.google.com/store/apps/category/GAME_TRIVIA",
		"https://play.google.com/store/apps/category/GAME_WORD",
		"https://play.google.com/store/apps/category/FAMILY",
		"https://play.google.com/store/apps/category/FAMILY?age=AGE_RANGE1",
		"https://play.google.com/store/apps/category/FAMILY?age=AGE_RANGE2",
		"https://play.google.com/store/apps/category/FAMILY?age=AGE_RANGE3",
		"https://play.google.com/store/apps/category/FAMILY_ACTION",
		"https://play.google.com/store/apps/category/FAMILY_BRAINGAMES",
		"https://play.google.com/store/apps/category/FAMILY_CREATE",
		"https://play.google.com/store/apps/category/FAMILY_EDUCATION",
		"https://play.google.com/store/apps/category/FAMILY_MUSICVIDEO",
		"https://play.google.com/store/apps/category/FAMILY_PRETEND",
	},
	IOS: {
		"https://itunes.apple.com/us/genre/ios-books/id6018?mt=8",
		"https://itunes.apple.com/us/genre/ios-business/id6000?mt=8",
		"https://itunes.apple.com/us/genre/ios-education/id6017?mt=8",
		"https://itunes.apple.com/us/genre/ios-entertainment/id6016?mt=8",
		"https://itunes.apple.com/us/genre/ios-finance/id6015?mt=8",
		"https://itunes.apple.com/us/genre/ios-food-drink/id6023?mt=8",
		"https://itunes.apple.com/us/genre/ios-games/id6014?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-action/id7001?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-adventure/id7002?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-arcade/id7003?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-board/id7004?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-card/id7005?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-casino/id7006?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-dice/id7007?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-educational/id7008?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-family/id7009?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-music/id7011?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-puzzle/id7012?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-racing/id7013?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-role-playing/id7014?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-simulation/id7015?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-sports/id7016?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-strategy/id7017?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-trivia/id7018?mt=8",
		"https://itunes.apple.com/us/genre/ios-games-word/id7019?mt=8",
		"https://itunes.apple.com/us/genre/ios-health-fitness/id6013?mt=8",
		"https://itunes.apple.com/us/genre/ios-lifestyle/id6012?mt=8",
		"https://itunes.apple.com/us/genre/ios-magazines-newspapers/id6021?mt=8",
		"https://itunes.apple.com/us/genre/ios-medical/id6020?mt=8",
		"https://itunes.apple.com/us/genre/ios-music/id6011?mt=8",
		"https://itunes.apple.com/us/genre/ios-navigation/id6010?mt=8",
		"https://itunes.apple.com/us/genre/ios-news/id6009?mt=8",
		"https://itunes.apple.com/us/genre/ios-photo-video/id6008?mt=8",
		"https://itunes.apple.com/us/genre/ios-productivity/id6007?mt=8",
		"https://itunes.apple.com/us/genre/ios-reference/id6006?mt=8",
		"https://itunes.apple.com/us/genre/ios-shopping/id6024?mt=8",
		"https://itunes.apple.com/us/genre/ios-social-networking/id6005?mt=8",
		"https://itunes.apple.com/us/genre/ios-sports/id6004?mt=8",
		"https://itunes.apple.com/us/genre/ios-travel/id6003?mt=8",
		"https://itunes.apple.com/us/genre/ios-utilities/id6002?mt=8",
		"https://itunes.apple.com/us/genre/ios-weather/id6001?mt=8",
	},
}

//init system main entry initializatoin
func init() {
	//uniqueness
	rand.Seed(time.Now().UnixNano())
	//init certs
	pool = x509.NewCertPool()
	pool.AppendCertsFromPEM(pemCerts)

	//in
	Formatters = map[string]Formatter{
		"LIST-CATEGORY-IOS":     {IOS, showCategory},
		"LIST-CATEGORY-ANDROID": {ANDROID, showCategory},
		"CATEGORY-IOS":          {IOS, showCategory},
		"CATEGORY-ANDROID":      {ANDROID, showCategory},
	}
	pAppsData = make(chan *App)
	//recovery
	initRecov()
	//loggers
	initLogger(os.Stdout, os.Stdout, os.Stderr)
	//http
	httpInit()

}

//initHanlder specific initializatoin
func initHanlder() {
	//evt
	initEnvParams()
}

//initRecov is for dumpIng segv in
func initRecov() {
	//might help u
	defer func() {
		recvr := recover()
		if recvr != nil {
			fmt.Println("MAIN-RECOV-INIT: ", recvr)
		}
	}()
}

//os.Stdout, os.Stdout, os.Stderr
func initLogger(i, w, e io.Writer) {
	//just in case
	if !pShowConsole {
		infoLog = makeLogger(i, pLogDir, "storemeta", "INFO: ")
		warnLog = makeLogger(w, pLogDir, "storemeta", "WARN: ")
		errorLog = makeLogger(e, pLogDir, "storemeta", "ERROR: ")
	} else {
		infoLog = log.New(i,
			"INFO: ",
			log.Ldate|log.Ltime|log.Lmicroseconds)
		warnLog = log.New(w,
			"WARN: ",
			log.Ldate|log.Ltime|log.Lshortfile)
		errorLog = log.New(e,
			"ERROR: ",
			log.Ldate|log.Ltime|log.Lshortfile)
	}
}

//initEnvParams enable all OS envt vars to reload internally
func initEnvParams() {
	//just in-case, over-write from ENV
	for k, v := range pEnvVars {
		if os.Getenv(k) != "" {
			*v = os.Getenv(k)
		}
	}
	flag.BoolVar(&pShowConsole, "debug", pShowConsole, usageShowConsole)
	flag.BoolVar(&pShowConsole, "d", pShowConsole, usageShowConsole+" (shorthand)")

	flag.StringVar(&pAndroidStoreID, "android", pAndroidStoreID, usageAndroidStore)
	flag.StringVar(&pAndroidStoreID, "a", pAndroidStoreID, usageAndroidStore+" (shorthand)")

	flag.StringVar(&pIOSStoreID, "ios", pIOSStoreID, usageIOSStore)
	flag.StringVar(&pIOSStoreID, "i", pIOSStoreID, usageIOSStore+" (shorthand)")

	flag.BoolVar(&pIOSList, "list-category-ios", pIOSList, usageIOSList)
	flag.BoolVar(&pIOSList, "li", pIOSList, usageIOSList+" (shorthand)")

	flag.BoolVar(&pAndroidList, "list-category-android", pAndroidList, usageAndroidList)
	flag.BoolVar(&pAndroidList, "la", pAndroidList, usageAndroidList+" (shorthand)")

	flag.StringVar(&pIOSCategory, "category-ios", pIOSCategory, usageIOSCategory)
	flag.StringVar(&pIOSCategory, "ci", pIOSCategory, usageIOSCategory+" (shorthand)")

	flag.StringVar(&pAndroidCategory, "category-android", pAndroidCategory, usageIOSCategory)
	flag.StringVar(&pAndroidCategory, "ca", pAndroidCategory, usageIOSCategory+" (shorthand)")

	flag.StringVar(&pPrintFormat, "print-format", pPrintFormat, usagePrintFormat)
	flag.StringVar(&pPrintFormat, "pf", pPrintFormat, usagePrintFormat+" (shorthand)")

	flag.BoolVar(&pHelp, "h", pHelp, "Show this help/how-to")

	flag.BoolVar(&pHTTPServe, "http", pHTTPServe, usageHTTP)
	flag.StringVar(&pHTTPPort, "port", pHTTPPort, usageHTTPPort)

	flag.BoolVar(&pTestingConsole, "dummy", pTestingConsole, usageTestingConsole)
	flag.Parse()

	//either 1 should be present
	if pHelp {
		showUsage()
	}

	if pIOSStoreID == "" && pAndroidStoreID == "" &&
		!pIOSList && !pAndroidList &&
		pIOSCategory == "" && pAndroidCategory == "" &&
		!pHTTPServe && !pTestingConsole {
		showUsage()
	}

	if pAndroidStoreID != "" || pIOSStoreID != "" {
		ands := strings.Split(pAndroidStoreID, ",")
		ioss := strings.Split(pIOSStoreID, ",")
		queryStoreIds(ands, ioss)
	}
}

//initHTTPRouters , init the routing
func initHTTPRouters() {
	pRouter := httprouter.New()
	pRouter.GET("/", indexHandler)
	pRouter.GET("/:mode", formatHandler)
	pRouter.POST("/:mode", formatHandler)
	pRouter.GET("/:mode/", formatHandler)
	pRouter.POST("/:mode/", formatHandler)
	fmt.Println("Storemeta\n\nVer: "+pVersion+"\n\nReady to serve @ port:", strings.Replace(pHTTPPort, ":", "", -1))
	log.Fatal(http.ListenAndServe(":"+pHTTPPort, pRouter))
}

//formatLogger try to init all filehandles for logs
func formatLogger(fdir, fname, pfx string) string {
	t := time.Now()
	r := regexp.MustCompile("[^a-zA-Z0-9]")
	p := t.Format("2006-01-02") + "-" + r.ReplaceAllString(strings.ToLower(pfx), "")
	s := path.Join(pLogDir, fdir)
	if _, err := os.Stat(s); os.IsNotExist(err) {
		//mkdir -p
		os.MkdirAll(s, os.ModePerm)
	}
	return path.Join(s, p+"-"+fname+".log")
}

//makeLogger initialize the logger either via file or console
func makeLogger(w io.Writer, ldir, fname, pfx string) *log.Logger {
	logFile := w
	if !pShowConsole {
		var err error
		logFile, err = os.OpenFile(formatLogger(ldir, fname, pfx), os.O_RDWR|os.O_CREATE|os.O_APPEND, 0664)
		if err != nil {
			log.Println(err)
		}
	}
	//give it
	return log.New(logFile,
		pfx,
		log.Ldate|log.Ltime|log.Lmicroseconds|log.Lshortfile)

}

//dumpW log into warning
func dumpW(s ...interface{}) {
	warnLog.Println(s...)
}

//dumpE log into error
func dumpE(s ...interface{}) {
	errorLog.Println(s...)
}

//dumpI log into info
func dumpI(s ...interface{}) {
	infoLog.Println(s...)
}

//Write override the log.print
func (w logOverride) Write(bytes []byte) (int, error) {
	//return fmt.Print(w.Prefix + time.Now().UTC().Format("2006-01-02 15:04:05.999") + " " + string(bytes))
	return fmt.Print(string(bytes))
}

//overrideLogger reset the log.print to customized
func overrideLogger(pfx string) {
	log.SetFlags(0)
	log.SetOutput(&logOverride{Prefix: pfx})
}

//showUsage
func showUsage() {

	msg := `



	Example:



		./storemeta -list-category-android

		./storemeta -list-category-ios

		./storemeta -category-android=GAME_ACTION

		./storemeta -category-ios=GAMES_ACTION

		./storemeta  -a="com.google.android.apps.photos"

		or

		./storemeta  -i="293622097"

		or

		./storemeta  -a="com.google.android.apps.photos" -i="293622097"

		or

		./storemeta  -a="com.google.android.apps.plus,com.google.android.launcher,com.sphero.sprk"

		or

		./storemeta  -i="544007664,535886823,643496868"

`

	fmt.Println()
	fmt.Println("Ver:", pVersion)
	fmt.Println()
	flag.PrintDefaults()
	fmt.Println(msg)
	os.Exit(0)
}
